numbers/Monte_Carlo.go

```diff
@@ -1,23 +1,104 @@
+//that divides the function support into discrete bins with optimized sampling heights.
+//maxSampleDepth triggers the log.Fatalf in the RejectionSample func, and samples is the number of values to be returned.
+func FastRejectionSampler(xLeft float64, xRight float64, f func(float64) float64, bins int, maxSampleDepth int, samples int, randSeed bool) []float64 {
+	if randSeed {
+		rand.Seed(time.Now().UnixNano())
+	}
+
+	var answer []float64 = make([]float64, samples)
+	if xLeft >= xRight {
+		log.Fatalf("Error in FastRejectionSample: xRight must be greater than xLeft.")
+	}
+
+	var support float64 = xRight - xLeft
+	var stepSize float64 = support / float64(bins)
+	var binHeights []float64 = make([]float64, bins)
+	var currLeft float64 = xLeft
+	var currRight float64 = xLeft + stepSize
+	var sumHeights float64
+
+	var fCurrLeft, fCurrRight float64
+	var firstTime bool = true
+
+	for i := 0; i < bins; i++ {
+		if firstTime {
+			firstTime = false
+			fCurrLeft = f(currLeft)
+			fCurrRight = f(currRight)
+			binHeights[i] = common.MaxFloat64(fCurrLeft, fCurrRight)
+		} else {
+			fCurrLeft = fCurrRight
+			currRight += stepSize
+			fCurrRight = f(currRight)
+			binHeights[i] = common.MaxFloat64(f(currLeft), f(currRight))
+		}
+		sumHeights += binHeights[i]
+	}
+	for j := 0; j < samples; j++ {
+		answer[j] = RejectionSampleChooseBin(currLeft, currRight, stepSize, f, maxSampleDepth, sumHeights, binHeights)
+	}
+	return answer
+}
+
+func RejectionSampleChooseBin(xLeft float64, xRight float64, stepSize float64, f func (float64) float64, maxIteration int, sumHeights float64, binHeights []float64) float64 {
+	var x, y float64
+	var currBin int
+	var currLeft, currRight float64
+	for i := 0; i < maxIteration; i++ {
+		currBin = chooseBin(sumHeights, binHeights)
+		currLeft = xLeft + float64(currBin)*stepSize
+		currRight = currLeft + stepSize
+		x = RandFloatInRange(currLeft, currRight)
+		y = f(x)
+		if RandFloatInRange(0.0, binHeights[currBin]) < y {
+			return x
+		}
+	}
+	log.Fatalf("Exceeded max iteration in RejectionSampleChooseBin.")
+	return -1.0
+} 
+
+//chooseBin picks which bin should be used for the FastRejectionSampler, where the choice of bin is weighted by its relative contribution to the overall integral of f.
+func chooseBin(sumHeights float64, binHeights []float64) int {
+	var rand float64 = rand.Float64()
+	var cumulative float64 = 0.0
+	for i := 0; i < len(binHeights); i++ {
+		cumulative += (binHeights[i] / sumHeights)
+		if cumulative > rand {
+			return i
+		}
+	}
+	log.Fatalf("Error in chooseBin: failed to choose a bin.")
+	return -1
+}
+
 //RejectionSample returns simulated values from an arbitrary function between a specified left and right bound using a simple rejection sampling method.
-func RejectionSample(xleft float64, xright float64, f func(float64) float64, maxIteration int) float64 {
+func RejectionSample(xLeft float64, xRight float64, yMax float64, f func(float64) float64, maxIteration int, returnX bool) float64 {
```

Comment:
Another option would be for the function to return an (x, y) pair.  If the calling function only wanted one they could use an underscore to get rid of the other.  This would also let calling functions get both if they ever wanted that.